src/main/scala/cosas/fns.scala

```diff
@@ -47,4 +47,156 @@ object fns {
 
   trait Fn3[A, B, C] extends AnyFn3 { type In1 = A; type In2 = B; type In3 = C }
 
+
+
+  /* Dependent functions aka dependent products */
+  trait AnyDepFn { type Out }
+  trait AnyDepFn0 extends AnyDepFn
+  trait AnyDepFn1 extends AnyDepFn { type In1 }
+  trait DepFn1[I,O] extends AnyDepFn1 { type In1 = I; type Out = O }
+  trait AnyDepFn2 extends AnyDepFn { type In1; type In2 }
+  trait DepFn2[I1,I2,O] extends AnyDepFn2 { type In1 = I1; type In2 = I2; type Out = O }
+  // TODO moar arities
+  // trait AnyDepFn3 extends AnyDepFn { type In1; type In2 }
+
+  trait AnyDepFn1Composition extends AnyDepFn1 {
+
+    type First <: AnyDepFn1
+    val first: First
+    type Second <: AnyDepFn1 { type In1 >: First#Out }
+    val second: Second
+
+    type In1 = First#In1
+    type Out = Second#Out
+  }
+
+  class Composition[
+    F <: AnyDepFn1,
+    S <: AnyDepFn1 { type In1 >: F#Out }
+  ]
+  (val first: F, val second: S) extends AnyDepFn1Composition {
+
+    type First = F
+    type Second = S
+
+    // implicit def appForMe[
+    //   X1 <: F#In1,
+    //   M <: F#Out,
+    //   O <: S#Out
+    // ](implicit
+    //   appF: App1[F,X1,M],
+    //   appS: App1[S,M,O]
+    // ): App1[this.type,X1,O] = App1 { x1: X1 => appS(appF(x1)) }
+  }
+
+  implicit final def depFn1Ops[DF <: AnyDepFn1, I0 <: DF#In1, O0 <: DF#Out](df: DF): DepFn1Ops[DF,I0,O0] =
+    DepFn1Ops(df)
+  case class DepFn1Ops[DF <: AnyDepFn1, I <: DF#In1, O <: DF#Out](val df: DF) extends AnyVal {
+
+    def at(f: I => O): App1[DF,I,O] =
+      App1(f)
+  }
```

Comment:
I think that `DepFn2Ops` is more correct than this. Are there any problems with doing it like this?

```scala
implicit def depFn1Ops[DF <: AnyDepFn1](df: DF): DepFn1Ops[DF] = DepFn1Ops(df)

case class DepFn1Ops[DF <: AnyDepFn1](val df: DF) extends AnyVal {
  def at[I <: DF#In1, O <: DF#Out](f: I => O): App1[DF, I, O] = App1(f)
}
```